src/devices/src/virtio/balloon/device.rs

```diff
@@ -0,0 +1,920 @@
+        }
+
+        // Compact pages into ranges.
+        let page_ranges = compact_page_frame_numbers(&mut pages);
+
+        // Remove the page ranges.
+        for (page_frame_number, range_len) in page_ranges {
+            let guest_addr = GuestAddress((page_frame_number as u64) << VIRTIO_BALLOON_PFN_SHIFT);
+
+            match remove_range(
+                &mem,
+                (guest_addr, u64::from(range_len) << VIRTIO_BALLOON_PFN_SHIFT),
+            ) {
+                Ok(_) => continue,
+                Err(e) => {
+                    error!("Error removing memory range: {:?}", e);
+                }
+            };
+        }
+
+        Ok(())
+    }
+
+    pub(crate) fn process_deflate_queue(&mut self) -> bool {
+        let mem = match self.device_state {
+            DeviceState::Activated(ref mem) => mem,
+            // This should never happen, it's been already validated in the event handler.
+            DeviceState::Inactive => unreachable!(),
+        };
+        METRICS.balloon.deflate_count.inc();
+
+        let queue = &mut self.queues[DEFLATE_INDEX];
+        let mut needs_interrupt = false;
+
+        while let Some(head) = queue.pop(&mem) {
+            queue.add_used(&mem, head.index, 0);
+            needs_interrupt = true;
+        }
+
+        needs_interrupt
+    }
+
+    pub(crate) fn process_stats_queue(&mut self) -> std::result::Result<(), DeviceError> {
+        let mem = match self.device_state {
+            DeviceState::Activated(ref mem) => mem,
+            // This should never happen, it's been already validated in the event handler.
+            DeviceState::Inactive => unreachable!(),
+        };
+        METRICS.balloon.stats_updates_count.inc();
+
+        while let Some(head) = self.queues[STATS_INDEX].pop(&mem) {
+            if let Some(prev_stats_desc) = self.stats_desc_index {
+                // We shouldn't ever have an extra buffer if the driver follows
+                // the protocol, but return it if we find one.
+                error!("balloon: driver is not compliant, more than one stats buffer received");
+                self.queues[STATS_INDEX].add_used(&mem, prev_stats_desc, 0);
+            }
+            for index in (0..head.len).step_by(SIZE_OF_STAT) {
+                // Read the address at position `index`. The only case
+                // in which this fails is if there is overflow,
+                // in which case this descriptor is malformed,
+                // so we ignore the rest of it.
+                let addr = head
+                    .addr
+                    .checked_add(index as u64)
+                    .ok_or(DeviceError::MalformedDescriptor)?;
+                let stat = mem
+                    .read_obj::<BalloonStat>(addr)
+                    .map_err(|_| DeviceError::MalformedDescriptor)?;
+                self.latest_stats
+                    .update_with_stat(&stat)
+                    .map_err(|_| DeviceError::MalformedPayload)?;
```

Comment:
No metric in case of stats update fail ?

Reply:
Added a metric in case of stats update failure.